### DISCUSSION TITLE
PnP.Framework Approach to SharePoint REST API Throttling

### DISCUSSION THREAD

**Discussion by anthonywhite** [pnp/pnpcore#794] [Q&A] [answered]
Does the `PnP.Framework.Provisioning` engine follow all the recommendations from Microsoft to avoid/detect throttling on the SharePoint REST API?

ref [Avoid getting throttled or blocked in SharePoint Online](https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online)

If we occasionally get e.g. `System.Net.Sockets.SocketException` from a call to `ApplyProvisioningTemplate`, could there be an underlying throttling issue?

**Comment by jansenbe** [comment]
PnP Framework does apply the throttling guidance for both CSOM as REST calls @anthonywhite. SocketExceptions are also retried, see https://github.com/pnp/pnpframework/blob/dev/src/lib/PnP.Framework/Http/RetryHandler.cs for the used code.

**Reply by anthonywhite** [reply]
Thanks @jansenbe (thought it would!). A few more questions:

Presumably our `SocketExceptions` are simply because failures continued beyond the default no. of retries?

Is there a way to override the default no. of retries or time interval being used by `ApplyProvisioningTemplate`? (I looked at `RetryHandler` -> `PnPHttpClient` but could not see how to set this)

PS Is [this](https://docs.microsoft.com/en-us/dotnet/api/microsoft.sharepoint.client.webextensions.applyprovisioningtemplate?view=sharepoint-pnpcoreol-3.2.1810) still the correct link for docs on `PnP.Framework` (can't find it on GitHub like `PnP.Core` ?)

**Reply by jansenbe** [reply]
@anthonywhite : there's limited documentation for PnP Framework across various blog posts, only PnP Core SDK has a centralized documentation site. The socket exceptions should not be the result of the number of retries running out, you'll get a different exception when that happens. For the parts using PnP Core SDK you can get notified about throttling events, see https://pnp.github.io/pnpcore/using-the-sdk/basics-eventhub.html#using-the-eventhub for more context. At the moment we don't have such a thing for PnP Framework however.

**Reply by anthonywhite** [reply]
> The socket exceptions should not be the result of the number of retries running out, you'll get a different exception when that happens. 

Is that assuming we are using 1.8.0 as @patrikhellgren mentioned?

**Comment by patrikhellgren** [comment, marked as answer]
@anthonywhite What version of PnP.Framework are you using? Please make sure you are using version 1.8 or later since that was the first version with the handling of SocketExceptions.

**Reply by anthonywhite** [reply]
Ah thanks @patrikhellgren , yes we still have 1.7.0 in our build 👍🏻
